internal/state/resolver/resolver.go

```diff
@@ -0,0 +1,153 @@
+func (e *ServiceResolverImpl) Resolve(ctx context.Context, svc *v1.Service, svcPort int32) ([]Endpoint, error) {
+	if svc == nil {
+		return nil, fmt.Errorf("cannot resolve a nil Service")
+	}
+
+	// We list EndpointSlices using the Service Name Index Field we added as an index to the EndpointSlice cache.
+	// This allows us to perform a quick lookup of all EndpointSlices for a Service.
+	var endpointSliceList discoveryV1.EndpointSliceList
+	err := e.client.List(
+		ctx,
+		&endpointSliceList,
+		client.MatchingFields{sdk.KubernetesServiceNameIndexField: svc.Name},
+		client.InNamespace(svc.Namespace),
+	)
+
+	if err != nil || len(endpointSliceList.Items) == 0 {
+		return nil, fmt.Errorf("no endpoints found for Service %s", client.ObjectKeyFromObject(svc))
+	}
+
+	return resolveEndpoints(svc, svcPort, endpointSliceList)
+}
+
+func resolveEndpoints(svc *v1.Service, svcPort int32, endpointSliceList discoveryV1.EndpointSliceList) ([]Endpoint, error) {
+	targetPort, err := getTargetPort(svc, svcPort)
+	if err != nil {
+		return nil, err
+	}
+
+	capacity := calculateEndpointSliceCapacity(endpointSliceList.Items, targetPort)
+
+	if capacity == 0 {
+		return nil, fmt.Errorf("no valid endpoints found for Service %s and port %d", client.ObjectKeyFromObject(svc), svcPort)
+	}
+
+	endpoints := make([]Endpoint, 0, capacity)
+
+	for _, eps := range endpointSliceList.Items {
+
+		if ignoreEndpointSlice(eps, targetPort) {
+			continue
+		}
+
+		for _, endpoint := range eps.Endpoints {
+
+			if !endpointReady(endpoint) {
+				continue
+			}
+
+			for _, address := range endpoint.Addresses {
+				ep := Endpoint{Address: address, Port: targetPort}
+				endpoints = append(endpoints, ep)
+			}
+		}
+	}
+
+	return endpoints, nil
+}
+
+func getTargetPort(svc *v1.Service, svcPort int32) (int32, error) {
+	for _, port := range svc.Spec.Ports {
+		if port.Port == svcPort {
+			val := port.TargetPort.IntValue()
+			if val == 0 {
+				break
+			}
+			return int32(val), nil
+		}
+	}
+
+	return 0, fmt.Errorf("no matching target port for Service %s/%s and port %d", svc.Namespace, svc.Name, svcPort)
+}
+
+func ignoreEndpointSlice(endpointSlice discoveryV1.EndpointSlice, targetPort int32) bool {
+	return endpointSlice.AddressType != discoveryV1.AddressTypeIPv4 || !targetPortExists(endpointSlice.Ports, targetPort)
+}
+
+func calculateEndpointSliceCapacity(endpointSlices []discoveryV1.EndpointSlice, targetPort int32) (capacity int) {
+	for _, es := range endpointSlices {
+
+		if ignoreEndpointSlice(es, targetPort) {
+			continue
+		}
+
+		for _, e := range es.Endpoints {
+			if !endpointReady(e) {
+				continue
+			}
+			capacity += len(e.Addresses)
+		}
+	}
+
+	return
+}
+
+func endpointReady(endpoint discoveryV1.Endpoint) bool {
+	ready := endpoint.Conditions.Ready
+	return ready != nil && *ready
+}
+
+func targetPortExists(ports []discoveryV1.EndpointPort, targetPort int32) bool {
+	for _, port := range ports {
+		if port.Port == nil {
```

Comment:
"When a port is defined with a nil port value, it indicates "all ports"."

Does anyone know what Service definitions would cause an "all ports" endpoint configuration? I read all ports as indicating either this should return `true` or we test and document special cases where we don't support a Service config.

Reply:
The only way I can produce an EndpointSlice with nil ports is by creating a Service that has no endpoints:
```
addressType: IPv4
apiVersion: discovery.k8s.io/v1
endpoints: null
kind: EndpointSlice
metadata:
  creationTimestamp: "2022-09-20T22:51:50Z"
  generateName: target-svc-
  generation: 1
  labels:
    endpointslice.kubernetes.io/managed-by: endpointslice-controller.k8s.io
    kubernetes.io/service-name: target-svc
  name: target-svc-pxhxr
  namespace: default
  ownerReferences:
  - apiVersion: v1
    blockOwnerDeletion: true
    controller: true
    kind: Service
    name: target-svc
    uid: 903364aa-a01e-4087-b72b-f0229cf2f84b
  resourceVersion: "2897033"
  uid: 13571449-21d2-4553-8be1-774829204869
ports: null
```

Reply:
nvm that's a different case -- `When ports is empty, it indicates that there are no defined ports.`

Reply:
I'm not sure what Service definition would cause an "all ports" endpoint config. 

`Service.Spec.Ports` is a required field as is `Service.Spec.Ports.Port`

Reply:
> I read all ports as indicating either this should return true 

Agreed. I'll fix this to return true.

Reply:
Asked about this in the Kubernetes slack: https://kubernetes.slack.com/archives/C09NXKJKA/p1663971959776179

Reply:
Haven't heard back from the k8s owners, but after reading the endpoint slice code it looks like the only way this can happen is if a user manually creates an endpoint slice object.

Reply:
`By default, the control plane creates and manages EndpointSlices to have no more than 100 endpoints each. You can configure this with the --max-endpoints-per-slice [kube-controller-manager](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/) flag, up to a maximum of 1000.`
https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/